[Merged by Bors] - feat(ring_theory/dedekind_domain): Chinese remainder theorem for Dedekind domains

Vierkantor opened the pull request:
The general Chinese remainder theorem states `R / I` is isomorphic to a product of `R / (P i ^ e i)`, where `P i` are comaximal, i.e.  `P i + P j = 1` for `i ≠ j`, and the infimum of all `P i` is `I`.

In a Dedekind domain the theorem can be stated in a more friendly way, namely that the `P i` are the factors (in the sense of a unique factorization domain) of `I`. This PR provides two ways of doing so, and includes some more lemmas on the ideals in a Dedekind domain.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #13066
- [x] depends on: #13161 (oops, forgot to cherry-pick this one in the original version of this PR!)

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(ring_theory/dedekind_domain): Chinese remainder theorem for Dedekind domains

The general Chinese remainder theorem states `R / I` is isomorphic to a product of `R / (P i ^ e i)`, where `P i` are comaximal, i.e.  `P i + P j = 1` for `i ≠ j`, and the infimum of all `P i` is `I`.

In a Dedekind domain the theorem can be stated in a more friendly way, namely that the `P i` are the factors (in the sense of a unique factorization domain) of `I`.

Vierkantor committed:
Fix usage of renamed lemma

riccardobrasca commented on src/ring_theory/dedekind_domain/ideal.lean:
```diff
@@ -855,3 +855,165 @@ end
+  P ≤ Q ↔ P = Q :=
+⟨(h P hP0 hP).eq_of_le hQ.ne_top, eq.le⟩
+
+variables {R} [is_domain R]
```
You can move this line after `ideal.coprime_of_no_prime_ge`, that doesn't need anything different. Also, `R` is already implicit.

riccardobrasca commented on src/ring_theory/dedekind_domain/ideal.lean:
```diff
@@ -855,3 +855,165 @@ end
+  I^n ≤ P ↔ I ≤ P :=
+⟨ideal.le_of_pow_le_prime, λ h, trans (ideal.pow_le_self hn) h⟩
+
+lemma ideal.prod_le_prime {ι : Type*} {s : finset ι} {f : ι → ideal R} {P : ideal R}
```
Maybe `P` can be explicit here? But I am not sure.

Vierkantor replied:
I'll leave it implicit since it occurs on both sides of the `↔` below.

riccardobrasca commented on src/ring_theory/dedekind_domain/ideal.lean:
```diff
@@ -855,3 +855,165 @@ end
+prime powers. -/
+lemma is_dedekind_domain.inf_prime_pow_eq_prod {ι : Type*}
+  (s : finset ι) (f : ι → ideal R) (e : ι → ℕ)
+  (prime : ∀ i, prime (f i)) (coprime : ∀ i j, i ≠ j → f i ≠ f j) :
```
The two assumptions are needed only for `i ∈ s`, right?

riccardobrasca commented on src/ring_theory/dedekind_domain/ideal.lean:
```diff
@@ -855,3 +855,165 @@ end
+
+/-- **Chinese remainder theorem** for a Dedekind domain: `R ⧸ I` factors as `Π i, R ⧸ (P i ^ e i)`,
+where `P i` ranges over the prime factors of `I` and `e i` over the multiplicities. -/
+noncomputable def is_dedekind_domain.quotient_equiv_pi_factors {I : ideal R} (hI : I ≠ ⊥) :
```
Is this really false for `I = ⊥`? I am maybe confused, but the product would be over the empty set, so it seems to work.

Vierkantor replied:
The quotient by ⊥ is the whole ring `R` and the product over the empty set would be the trivial ring, right? So I don't think this holds for `I = ⊥`.

riccardobrasca replied:
You're right, sorry.

Vierkantor committed:
Simplify `variables` declarations

Vierkantor committed:
Slightly generalize `inf_prime_pow_eq_prod` (the conditions only need to hold for `i ∈ s`)

The code is slightly more awkward now because we have to pass around membership in a few places. But it works fine.
